test/performance/odp_ipsecfwd.c

```diff
@@ -0,0 +1,1307 @@
+	       "                      values whitespace separated:\n"
+	       "\n"
+	       "                          <line in file> DstIPv4 SubNetMask NetIf DstMac\n"
+	       "\n"
+	       "                      DstIPv4 and SubNetMask define a matchable prefix and NetIf\n"
+	       "                      and DstMac define the outgoing interface and destination\n"
+	       "                      MAC address for a match. NetIf should be one of the\n"
+	       "                      interfaces passed with \"--interfaces\" option\n"
+	       "  -h, --help          This help.\n"
+	       "\n");
+}
+
+static odp_bool_t setup_ipsec(prog_config_t *config)
+{
+	odp_queue_param_t q_param;
+	odp_ipsec_config_t ipsec_config;
+
+	odp_queue_param_init(&q_param);
+	q_param.type = ODP_QUEUE_TYPE_SCHED;
+	q_param.sched.prio = odp_schedule_default_prio();
+	q_param.sched.sync = ODP_SCHED_SYNC_PARALLEL;
+	q_param.sched.group = ODP_SCHED_GROUP_ALL;
+	config->compl_q = odp_queue_create(PROG_NAME, &q_param);
+
+	if (config->compl_q == ODP_QUEUE_INVALID) {
+		ODPH_ERR("Error creating IPsec completion queue\n");
+		return false;
+	}
+
+	odp_ipsec_config_init(&ipsec_config);
+	ipsec_config.inbound_mode = ODP_IPSEC_OP_MODE_ASYNC;
+	ipsec_config.outbound_mode = ODP_IPSEC_OP_MODE_ASYNC;
+	ipsec_config.inbound.default_queue = config->compl_q;
+	/* For tunnel to tunnel, we need to parse up to this to check the UDP port for SA. */
+	ipsec_config.inbound.parse_level = ODP_PROTO_LAYER_L4;
+
+	if (odp_ipsec_config(&ipsec_config) < 0) {
+		ODPH_ERR("Error configuring IPsec\n");
+		return false;
+	}
+
+	return true;
+}
+
+static odp_bool_t check_alg_support(int cipher_idx)
+{
+	const size_t len = sizeof(exposed_algs) / sizeof(exposed_algs[0]);
+
+	for (size_t i = 0U; i < len; ++i)
+		if (exposed_algs[i].idx == cipher_idx)
+			return true;
+
+	return false;
+}
+
+static void create_sa_entry(uint32_t dir, uint32_t spi, const char *src_ip_str,
+			    const char *dst_ip_str, int cipher_idx, uint8_t *cipher_key,
+			    uint8_t *cipher_key_extra, int auth_idx, uint8_t *auth_key,
+			    uint8_t *auth_key_extra, uint32_t icv_len, uint32_t ar_ws,
+			    prog_config_t *config)
+{
+	uint32_t src_ip, dst_ip;
+	odp_queue_param_t q_param;
+	odp_ipsec_sa_param_t sa_param;
+	odp_ipsec_crypto_param_t crypto_param;
+	odp_ipsec_sa_t sa;
+
+	if (config->num_sas == MAX_SAS) {
+		ODPH_ERR("Maximum number of SAs parsed (%u), ignoring rest\n", MAX_SAS);
+		return;
+	}
+
+	if (!check_alg_support(cipher_idx)) {
+		ODPH_ERR("Unsupported cipher/authentication algorithm for SA %u\n", spi);
+		return;
+	}
+
+	if (odph_ipv4_addr_parse(&src_ip, src_ip_str) < 0 ||
+	    odph_ipv4_addr_parse(&dst_ip, dst_ip_str) < 0) {
+		ODPH_ERR("Error parsing IP addresses for SA %u\n", spi);
+		return;
+	}
+
+	if (spi > UINT16_MAX) {
+		ODPH_ERR("Invalid SPI value for SA %u\n", spi);
+		return;
+	}
+
+	if (spi_to_sa_map[dir][spi] != NULL) {
+		ODPH_ERR("Duplicate SA config %u\n", spi);
```

Comment:
Maybe "non-unique SPIs not supported" as in general multiple outbound SAs with the same SPI are ok?